### ISSUE TITLE
File download not possible in IE11 (Internet Explorer) - VeliovGroup/Meteor-Cookies problem?

### ISSUE THREAD

**Issue by derwok** [VeliovGroup/Meteor-Files#327]
**meteor 1.4.2.3
ostrio:cookies@2.0.5
ostrio:files@1.7.6**

Unfortunately we have to support IE11  :-( (I know, I know... - sad world).
VeliovGroup/Meteor-Files works great with all other browsers - so we where hit by surprise when we found out the infamous IE11 makes trouble again, here...  :-(

So what happens?
With IE11 it is possible to upload and remove files. But it is not possible to download files with an URL like this, while having a downloadCallback that checks for a logged in user:
http://www.myserver.com:3100/cdn/storage/AttachmentsCollection/oDYpTaG7k7xQGNEW3/original/oDYpTaG7k7xQGNEW3.txt?download=true

** IE11 gives a "HTTP 404" error.**

We have to protect files against not logged in users. Looking on the server the 404 problem seems to be as our `FilesCollection/downloadCallback` returns a "false" here:

        if (! this.userId) {
            console.log("Attachment download prohibited. User not logged in.");
            return false;
        }

If we dump the "this" in the above context we see that there is actually no `userId` available:

    [...]
    I20170113-07:56:19.481(1)?   params:
    I20170113-07:56:19.483(1)?    { query: { download: 'true' },
    I20170113-07:56:19.484(1)?      _id: 'oDYpTaG7k7xQGNEW3',
    I20170113-07:56:19.484(1)?      version: 'original',
    I20170113-07:56:19.484(1)?      name: 'oDYpTaG7k7xQGNEW3.txt?download=true' },
    I20170113-07:56:19.494(1)?   user: [Function: user],
    I20170113-07:56:19.494(1)?   userId: null }

Taking a closer look at the server log we see that the problem seems to arise way before the actual download. If the IE11 user signs in to our app with the meteor accounts package we see this in the error log:

```
W20170113-07:37:52.782(1)? (STDERR) Error: `value` required in setHeader("Set-Cookie", value).
W20170113-07:37:52.782(1)? (STDERR)     at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:342:11)
W20170113-07:37:52.782(1)? (STDERR)     at ServerResponse.appendHeader (/Users/wok/.meteor/packages/webapp/.1.3.12.10jgpt3++os+web.browser+web.cordova/npm/node_modules/connect/lib/patch.js:95:33)
W20170113-07:37:52.782(1)? (STDERR)     at ServerResponse.res.setHeader (/Users/wok/.meteor/packages/webapp/.1.3.12.10jgpt3++os+web.browser+web.cordova/npm/node_modules/connect/lib/patch.js:126:19)
W20170113-07:37:52.783(1)? (STDERR)     at Object.handle (packages/ostrio_cookies.js:509:21)
W20170113-07:37:52.784(1)? (STDERR)     at next (/Users/wok/.meteor/packages/webapp/.1.3.12.10jgpt3++os+web.browser+web.cordova/npm/node_modules/connect/lib/proto.js:174:15)
W20170113-07:37:52.784(1)? (STDERR)     at Function.app.handle (/Users/wok/.meteor/packages/webapp/.1.3.12.10jgpt3++os+web.browser+web.cordova/npm/node_modules/connect/lib/proto.js:182:3)
W20170113-07:37:52.784(1)? (STDERR)     at Object.fn [as handle] (/Users/wok/.meteor/packages/webapp/.1.3.12.10jgpt3++os+web.browser+web.cordova/npm/node_modules/connect/lib/proto.js:79:14)
W20170113-07:37:52.785(1)? (STDERR)     at next (/Users/wok/.meteor/packages/webapp/.1.3.12.10jgpt3++os+web.browser+web.cordova/npm/node_modules/connect/lib/proto.js:174:15)
W20170113-07:37:52.785(1)? (STDERR)     at Object.WebAppInternals.staticFilesMiddleware (packages/webapp/webapp_server.js:369:5)
W20170113-07:37:52.785(1)? (STDERR)     at packages/webapp/webapp_server.js:642:22
```
In the context of meteors "staticFilesMiddleware" here this line of interest: `at Object.handle (packages/ostrio_cookies.js:509:21)`
So, the root cause seems to be a problem with **VeliovGroup/Meteor-Cookies** here.
(By the way: should we file the issue over there?!)

Would be great if you could have look into this issue?

Thanks for all your great work!
We hobe you have a clue what's going on here...

Thanks in advance.
Wolfram.

**Comment by derwok** [comment]
Added version numbers at top of issue text...

**Comment by dr-dimitru** [comment]
Hello @derwok 

So, issue is in login/authentication process, not in download, right?

 1. What cookies do you have?
 2. What in localStorage?
 3. What value of `Meteor.connection._lastSessionId` in browser?
 4. Is user 100% logged in? `Meteor.user()`
 5. What limits of IE 11? Does it supports WebSockets?

**Comment by derwok** [comment]
Hi @dr-dimitru ,
Thanks for asking.

Yes - the problem seems to be during the login phase. But it seems to be in the code of your Meteor-Cookies sub-package, as the above call stack shows.
see line `at Object.handle (packages/ostrio_cookies.js:509:21)`.

Funny: the complete rest of the application with all Meteor.user() Meteor.userId checks works 100% perfect... So - metoer login  and session handling works and all checks in frontend and backend work also correctly. So the problem seems to be during login phase where your code tries to store the cookie for Meteor Files...
To your questions:
IE11 supports WebSockets according to here: http://caniuse.com/#feat=websockets
The user is completely logged in when he clicks on the download link (which does not work). At the moment when the above exception occurs - I have no clue, if the user is 100% logged in, because the exception call stack is somewhere beyond our control (meaning: not our code).

The problem with question 1-4 above is: The exception happens on server **during** login. So, as your questions are all client/browser questions: you want me to get this information client side **after** login - right? At this time point the server exception already happended. So, its too late...

I will try to collect the data soon. And comment below.

In the mean time - if you can get your hands on a Windows with IE11 on it somewhere... you can either:

1. Head over to https://www.4minitz.com
1. Login with demo/demo
1. Click on the meeting series "Test - Meteor-Files"
1. Click the date of today's meeting 2017-01-13
1. Scroll to attachments section and try to download the file

Or - if you may investigate client and server side:
```
git clone https://github.com/4minitz/4minitz.git
cd 4minitz
meteor run --settings ./settings_sample.json
```

And then login with IE11 browser and user: **demo** and password: **demo**.
You will immediately see the above server side exception on login.
And its not your code somewhere in the exception call stack...

**Comment by dr-dimitru** [comment]
@derwok Thank you for details.
I need to find a PC or deploy VM to investigate this.

Please provide Windows version and IE 11 full version including build (if exists).

As investigation may took some time, here is temporary solution - https://github.com/thomasspiesser/ostrio-files-bug/pull/1/files
Where we attempted to fix similar issue

Please read this thread - https://github.com/VeliovGroup/Meteor-Files/issues/159
I believe it has same roots.

What you're having now is *Server* issue not client.
[Here](https://github.com/VeliovGroup/Meteor-Cookies/blob/master/cookies.coffee#L254) at [Cookies package](https://github.com/VeliovGroup/Meteor-Cookies) empty value is passed to header. That means it's not sent from client.

Crated issue on Cookies package - https://github.com/VeliovGroup/Meteor-Cookies/issues/6

**Comment by derwok** [comment]
Hi,

I just managed to build the smallest meteor app to reproduce the error...
I was not able to reproduce the problem with the Meteor-Cookies package alone. I needed to add the Meteor-Files package. So, iInstead of cloning our full blown 4minitz app you can also clone this repo:

[TO-BE-DONE] 

After `meteor create` I just added 3 packages:


When you cloned & run the cookiestest meteor app: 
Then open webapp in browser, and simple register a user with arbitrary email (no validation happening!) and password of at least 6 chars. As soon as the user is registered, the user is is logged in and you will see the server exception triggered by the cookies package

Here are some answers to your questions:

**0. Im am testing on**

```
   "Windows 7 Professional" with ServicePack 1 and current patches
   "Internet Explorer" v11.0.9600.18537
```

According to here:
https://www.netmarketshare.com/browser-market-share.aspx?qprid=2&qpcustomd=0

IE11 still holds a share of >10% of the global browser market. Though it is known for
handling some things not 100% as the rest of the big browsers... I think it is worth a
consideration in our apps, as especially (large) companies currently still stick to Windows 7 or Windows 8 and IE11 is the latest Internet Explorer that runs on these systems. In these companies staff is not allowed to install FF / Chrome and MS Edge wont run on these legacy windows versions. Edge need Windows 10.


**1.What cookies do you have?**

In IE11 you can show the cookies as descibed here:
http://stackoverflow.com/questions/19941029/how-do-i-view-cookies-in-internet-explorer-11-using-developer-tools

When I do as described above, and I reload my page and login, I see the server side exception. 
But there show no cookies up on the client. The windows that should show the cookies is simply empty.


**2.What in localStorage?**
After server side login exception, I am nevertheless logged in on the client.
When opening the F12 dev console, and I enter "localStorage" I get the following output:

```
{
   [functions]: ,
   __proto__: { },
   constructor: { },
   length: 3,
   Meteor.loginToken: "6XOUVJ1p-0nSR4YNv-zGJyyeQjZ79ZcP8EK0NjB1be4",
   Meteor.loginTokenExpires: "Fri Apr 14 2017 10:08:00 GMT+0200 (Mitteleuropäische Sommerzeit)",
   Meteor.userId: "ELJB4aou57pADMAAA",
   remainingSpace: 4999822
}
```


**3.What value of  Meteor.connection._lastSessionId  in browser?**

"34PBFFQRbfA8cMKqA"

**4.Is user 100% logged in?  Meteor.user()**
```
On the client:
{
   [functions]: ,
   __proto__: { },
   _id: "ELJB4aou57pADMAAA",
   emails: [ ]
}
```

**5.What limits of IE 11? Does it supports WebSockets?**

Yes IE11 is known to handle things a bit strange sometimes...
Nevertheless here is stated websockets and localStorage work just fine in IE11:
http://caniuse.com/#feat=websockets
http://caniuse.com/#search=localstorage

**Comment by dr-dimitru** [comment]
Hello @derwok ,

Could you try `v1.7.7` from dev branch? (*have no access to IE now*)

**Comment by derwok** [comment]
Hey! Great News! Yes, sure! I'll try! Stay tuned.

**Comment by dr-dimitru** [comment]
@derwok pushed more critical changes related to this issue into `dev`

**Comment by derwok** [comment]
Maybe this is of help?
http://stackoverflow.com/questions/22690114/internet-explorer-11-wont-set-cookies-on-a-site

**Comment by dr-dimitru** [comment]
@derwok thank you, helped a lot.

Updated cookies package to [`v2.1.2`](https://github.com/VeliovGroup/Meteor-Cookies/releases/tag/2.1.2)

Please see updated `v1.7.7` in `dev`
Let me know if this issue still persists for you.

**Comment by macrozone** [comment]
can also reproduce on IE11, update on 1.7.7 fixes the error about the cookies on file download,
but file can't be accessed. The mtok-cookie is set, but it's not send with the image-request. Not sure why this is happening

**Comment by macrozone** [comment]
ok here is the problem:

the cookie value has to be written in quotes:

document.cookie = 'x_mtok="abc1234"; path=/';

**Comment by dr-dimitru** [comment]
Hello @macrozone ,

> Btw. why re-implementing the whole cookie handling and not using an already existing cookie library that has been tested properly on all browsers to set and get cookies?

100% agree. Thats why [Meteor-Cookies](https://github.com/VeliovGroup/Meteor-Cookies) is based on top of [jshttp/cookie](https://github.com/jshttp/cookie) package.

I haven't found any reference, which tells to wrap cookie value into quotes. [jshttp/cookie](https://github.com/jshttp/cookie) package does not [wraps value](https://github.com/jshttp/cookie/blob/master/index.js#L119), and js-cookie package [leaves value unwrapped](https://github.com/js-cookie/js-cookie/blob/master/src/js.cookie.js#L92)

Could you please run [Meteor-Cookies](https://github.com/VeliovGroup/Meteor-Cookies) package tests? `meteor test-packages ./`

Need to confirm I'm subject of NIH syndrome... can't do anything about it... :(

**Comment by macrozone** [comment]
Hi @dr-dimitru 

Meteor-cookies does its own version of cookie-parsing (https://github.com/VeliovGroup/Meteor-Cookies/blob/master/cookies.js#L56) and setting (https://github.com/VeliovGroup/Meteor-Cookies/blob/master/cookies.js#L95). consider using an an npm package here

Yeah, That thing with wrapping the values in values in quotes is complete bs (its strange that it works...).

I just was very confused when debugging this on the IE. I tried to set an arbitrary cookie by pasting this in the IE11-console:

`document.cookie = 'thecookie=value; path=/';`

and checked the network tab on the IE to see if this cookie is present, and it wasnt.

Then i tried `document.cookie = 'thecookie="value; path=/';` and it worked...

But i  really don't trust the dev-tools in IE... sometimes cookies are there, sometimes not... its a mess

I forgot to write above that i wrapped the document.cookie in a `Meteor.setTimeout(() => {...}, 300)`

In your code, the timeout is 25. I did not investigate much on this, but maybe `Meteor.connection._lastSessionId` is not defined sometimes, maybe a timing issue?

**Comment by dr-dimitru** [comment]
I'm asking to run tests, as I got access to MS Win 8 IE 11. All Meteor-Cookies package tests is passed on bye end. Tests for protected files (*cookie validation*) is passed too. So, if you have a chance could you please run tests and/or try to access protect files at https://files.veliov.com

To do so:
 1. Go to https://files.veliov.com
 2. Sing up via Meteor Account / Twitter / Facebook / GitHub
 3. Upload a file
 4. Make sure it's protected (*lock icon*)
 5. Click on download button

If file is downloaded in IE - you're authorised via cookies

**Comment by macrozone** [comment]
I opend it on IE11 windows 8 and could not login, because the info button is above the login button...

I removed it know through dev tools... and yeah, it works in IE

**Comment by dr-dimitru** [comment]
dev and master currently at the same state.
files.veliov.com is up to date too

**Comment by dr-dimitru** [comment]
> because the info button is above the login button...

PR is welcome [here](https://github.com/VeliovGroup/Meteor-Files-Demos/blob/master/demo/client/styles/styles.sass#L393)

**Comment by macrozone** [comment]
oh i see... i think i checked out master not develop before, my bad!

**Comment by macrozone** [comment]
Ok i checked it out again and it still did not work. But i think i found now the real problem:

The cookie path was not set so i guess ie has here a different behavior.

I sent a PR: https://github.com/VeliovGroup/Meteor-Files/pull/333

**Comment by dr-dimitru** [comment]
@macrozone 

[RFC6265](tools.ietf.org/html/rfc6265) does not indicate any of cookie attributes as required.

> Each cookie begins with a name-value-pair, followed by __zero__ or more attribute-value pairs.

Need to confirm IE can't set cookie without `path` attribute

**Comment by macrozone** [comment]
ie can set cookies without path.

if you omit path, it sets the path to the current path. In my case this was "/de".
Since the image-urls did not start with that path, IE did refuse to send a cookie there.

I am not sure, why this worked in the other browsers though.

path = "/" makes it available for the whole domain. I think this is exactly what you want here.

**Comment by dr-dimitru** [comment]
>        If the cookie-attribute-list contains an attribute with an
>        attribute-name of "Path", set the cookie's path to attribute-
>        value of the last attribute in the cookie-attribute-list with an
>        attribute-name of "Path".  Otherwise, set the cookie's path to
>        the default-path of the request-url.


What is `default-path` is not explained, or I didn't found an explanation.
So, I guess IE uses `request-url` here, and your PR is reasonable in this case. But I still would like to see some test results...

**Comment by derwok** [comment]
Hi @dr-dimitru, @macrozone!
I cloned in the change of @macrozone 's PR (138a0133ebcb7f828c92dec09829ce080ca45fee) - in the local packages dir - and what should I say?!  ;-)

**For the first time in my life I could see stupid IE11 perform some download**
![ie11_download](https://cloud.githubusercontent.com/assets/4393698/22183947/9a24613e-e0c9-11e6-8541-4df318cdfd81.png)
(Of course I first had to logout/login with IE11 before it worked.)

**HOORAY!** Thank you very much to both of you!

I then checked upload and download functionality with these browsers:
* Safari 10.0.2 - OK
* Chrome  55.0.2883.95 (64-bit) - OK
* FireFox 50.1.0 - OK
* IE11 11.0.9600 - Finally: OK!!!

Just to ~doublecheck~ out of curiosity: I removed the patched line of @macrozone of commit 138a0133ebcb7f828c92dec09829ce080ca45fee - so I went back to:
```
     cookie.set 'x_mtok', Meteor.connection._lastSessionId
```
With every browser I did a logout / login then download try:
* Safari 10.0.2 - **FAIL**
* Chrome  55.0.2883.95 (64-bit) - **FAIL**
* FireFox 50.1.0 - **FAIL**
* IE11 11.0.9600 - Still: OK!!!  - surprise!

So, yes: I would vote for @macrozone 's PR.
;-)

Thanks for taking care.

Wolfram.

**Comment by dr-dimitru** [comment]
Hi @derwok ,

Thank you for tests, with surprising results.
Looking forward do merge @macrozone 's PR, minor changes is requested.
